ambry-network/src/main/java/com.github.ambry.network/LocalRequestResponseChannel.java

```diff
@@ -0,0 +1,157 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link RequestResponseChannel} that buffers messages in local queues.
+ * This class enables the Frontend router to call AmbryRequest methods in process.
+ */
+public class LocalRequestResponseChannel implements RequestResponseChannel {
+
+  private static final Logger logger = LoggerFactory.getLogger(LocalRequestResponseChannel.class);
+  private BlockingQueue<Request> requestQueue = new LinkedBlockingQueue<>();
+  private Map<Integer, List<ResponseInfo>> responseMap = new ConcurrentHashMap<>();
+  // buffer to hold size header that we strip off payloads
+  private static final byte[] sizeByteArray = new byte[Long.BYTES];
+
+  @Override
+  public void sendRequest(Request request) {
+    requestQueue.offer(request);
+    if (request instanceof LocalChannelRequest) {
+      LocalChannelRequest localRequest = (LocalChannelRequest) request;
+      logger.debug("Added request for {}, queue size now {}", localRequest.processorId, requestQueue.size());
+    }
+  }
+
+  @Override
+  public Request receiveRequest() throws InterruptedException {
+    Request request = requestQueue.take();
+    if (request instanceof LocalChannelRequest) {
+      LocalChannelRequest localRequest = (LocalChannelRequest) request;
+      logger.debug("Removed request for {}, queue size now {}", localRequest.processorId, requestQueue.size());
+    }
+    return request;
+  }
+
+  @Override
+  public void sendResponse(Send payloadToSend, Request originalRequest, ServerNetworkResponseMetrics metrics) {
+    try {
+      LocalChannelRequest localRequest = (LocalChannelRequest) originalRequest;
+      ResponseInfo responseInfo =
+          new ResponseInfo(localRequest.requestInfo, null, byteBufferFromPayload(payloadToSend));
+      List<ResponseInfo> responseList = getResponseList(localRequest.processorId);
+      synchronized (responseList) {
+        responseList.add(responseInfo);
+        logger.debug("Added response for {}, size now {}", localRequest.processorId, responseList.size());
+      }
+    } catch (IOException ex) {
+      logger.error("Could not extract response", ex);
+    }
+  }
+
+  /**
+   * Receive all queued responses corresponding to requests matching a processor id.
+   * @param processorId the processor id to match.
+   * @return the applicable responses.
+   */
+  public List<ResponseInfo> receiveResponses(int processorId) {
+    List<ResponseInfo> responseList = getResponseList(processorId);
+    synchronized (responseList) {
```

Comment:
maybe consider using a blocking queue instead of a plain ArrayList for `responseList` since it supports timed waits.

Reply:
Will do.